app/src/main/java/in/co/rajkumaar/amritarepo/timetable/AcademicTimetableActivity.java

```diff
@@ -1,303 +1,362 @@
+            savePref();
+            if (isNetworkAvailable()) {
+                TIMETABLE_URL = "https://intranet.cb.amrita.edu/TimeTable/PDF/";
+                buildTimetableUrl();
+                Intent intent = new Intent(AcademicTimetableActivity.this, WebViewActivity.class);
+                intent.putExtra("webview", TIMETABLE_URL);
+                intent.putExtra("zoom", true);
+                intent.putExtra("title", branch.getSelectedItem() + " " + batch.getSelectedItem() + " - Semester " + sem.getSelectedItem());
+                startActivity(intent);
+            } else {
+                Snackbar.make(view, "Device not connected to Internet", Snackbar.LENGTH_SHORT).show();
+            }
+        } else {
+            Snackbar.make(view, "Please select all the choices", Snackbar.LENGTH_SHORT).show();
+        }
+    }
+
+    public void downloadTimetable(View view) {
+        if (batch.getSelectedItemPosition() > 0 && branch.getSelectedItemPosition() > 0 && course.getSelectedItemPosition() > 0 && sem.getSelectedItemPosition() > 0 && year.getSelectedItemPosition() > 0) {
+            savePref();
+            if (ContextCompat.checkSelfPermission(AcademicTimetableActivity.this,
+                    Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                    != PackageManager.PERMISSION_GRANTED) {
+
+                ActivityCompat.requestPermissions(AcademicTimetableActivity.this,
+                        new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
+                        1);
+            } else {
+                if (isNetworkAvailable()) {
+                    String protocol = "https://";
+                    String intranet = getString(R.string.intranet);
+                    String timetable = getString(R.string.timetable);
+                    TIMETABLE_URL = protocol + intranet + timetable;
+                    buildTimetableUrl();
+                    new DownloadTask(AcademicTimetableActivity.this, TIMETABLE_URL, 0);
+                } else {
+                    Snackbar.make(view, "Device not connected to Internet.", Snackbar.LENGTH_SHORT).show();
+                }
+            }
+        } else {
+            Snackbar.make(view, "Please select all the choices", Snackbar.LENGTH_SHORT).show();
+        }
+    }
+
+    private void buildBranchesSpinner(int courseID) {
+        List<String> branchestemp = new ArrayList<>();
+        branchestemp.add("[Choose branch]");
+        if (!pref.contains(courses.get(courseID))) {
+            needupdatebranch = true;
+            getBranches(courseID, branchestemp);
+        } else {
+            branches = new ArrayList<>();
+            Gson gson = new Gson();
+            String json = pref.getString(courses.get(courseID), null);
+            Type listType = new TypeToken<ArrayList<String>>() {
+            }.getType();
+            ArrayList<String> parsedbranch = gson.fromJson(json, listType);
+            branches.addAll(parsedbranch);
```

Comment:
```suggestion
            branches.addAll(gson.fromJson(json, listType));
```
Avoid unnecessary variables.